### PR TITLE
feat: introduce new scheduling algorithm

### PR DESCRIPTION
**Pull request by shanicky**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://raw.githubusercontent.com/risingwavelabs/risingwave/17af8a747593ebdbfa826691daf75bdab7d14fa0/.github/contributor-license-agreement.txt).

## What's changed and what's your intention?

This PR attempts to introduce a new scheduling algorithm, primarily aimed at ensuring the affinity between vnodes and workers during multiple scheduling/rescheduling processes.

This PR attempts to use weighted rendezvous hashing for scheduling. [Rendezvous hashing](https://en.wikipedia.org/wiki/Rendezvous_hashing) is a algorithm similar to consistent hashing, used to assign items to specific workers.

Currently, our algorithm only guarantees affinity between vnodes and actors. This is no longer sufficient for the current scenarios. As more clusters utilize disk caches or other block caches, we need to ensure that vnodes maintain their affinity with workers across multiple reschedules as much as possible.

Furthermore, in our future plans to deprecate the meta store's actor table, a stable and evenly distributed algorithm that can guarantee affinity is essential. This PR aims to replace the existing simple scheduling algorithm with a more robust approach.

Currently, we have two extreme scenarios:

1. A large number of streaming jobs with many actors, but each streaming job has relatively low traffic. In this case, we can use a less uniform algorithm that ensures overall fairness based on the number of jobs, allowing us to focus more on maintaining affinity.  
2. Very few streaming jobs with few actors, but each job carries massive traffic. Under these circumstances, even slight imbalance can lead to significant traffic skew, so we need to ensure perfect uniformity in scheduling.

This PR specifically introduces a multi-level scheduling pattern involving actor -> vnode -> worker:

1. Assigns actors to workers using rendezvous hashing, distributed according to the workers' weights.  
2. Assigns vnodes to workers based on either the number of actors (acting as weight) or the original worker weights.  
3. Assigns vnodes to actors.

This PR still requires more testing to confirm corner cases and needs additional time to evaluate scheduling issues under different workloads.

### Allocation Matrix: `CapacityMode` × `BalancedBy`

| **Actor → Worker (`CapacityMode`) \ VNode → Worker (`BalancedBy`)** | **RawWorkerWeights**<br><sub>VNode weight = original worker weight</sub> | **ActorCounts**<br><sub>VNode weight = #actors on each worker</sub> |
| --- | --- | --- |
| **Weighted**<br><sub>Actors capped by worker weight</sub> | *Actor stage* strictly respects weights (e.g. W1:W2 = 1:3 → actors ≈ 1:3).<br>*VNode stage* again uses original weights (with a lower-bound of actor count).<br>Result: heavier workers get **more VNodes *per actor*** – machine-oriented load. | *Actor stage* still follows raw weights.<br>*VNode stage* uses actor counts as weights, so each actor gets roughly the **same number of VNodes**.<br>Good when you want per-actor fairness but still honor worker capacities. |
| **Unbounded**<br><sub>No actor cap; rendezvous hash only</sub> | *Actor stage* is weight-biased but unbounded – a light worker can still win actors via hashing.<br>*VNode stage* keeps raw weights, so a worker that “lucks out” with many actors might end up **CPU heavy but VNode light**. | *Actor stage* is free + biased.<br>*VNode stage* mirrors actor counts, giving near-equal VNodes per actor.<br>This combo yields the **fairest per-actor load** while letting hashing absorb cluster churn. |

#### Cheat-Sheet

* **CapacityMode::Weighted**  
  * Hard cap: actors ≤ worker weight (scale = 1).  
  * Highly predictable, mirrors hardware sizing.

* **CapacityMode::Unbounded**  
  * Actors limited only by rendezvous hash probability.  
  * Useful when worker capacity is unknown or elastic.

* **BalancedBy::RawWorkerWeights**  
  * VNodes skew toward heavier machines; each actor on a “big” worker handles more VNodes.  
  * Best when storage/IO throughput truly scales with weight.

* **BalancedBy::ActorCounts**  
  * VNodes proportional to actor counts, so every actor sees similar load.  
  * Ideal when **actor-level fairness** matters more than raw machine weight.

## Checklist

- [x] I have written necessary rustdoc comments.
- [x]  I have added necessary unit tests and integration tests.

### CODE REVIEW

File: src/meta/src/stream/stream_graph/assignment.rs [hunk 1]
```diff
@@ -0,0 +1,2127 @@
...
+    Weighted,
+
+    /// Ignore per-container quotas entirely—every container can take an unlimited number of items.
+    Unbounded,
+}
+
+/// Defines the vnode distribution strategy for hierarchical assignment.
+///
+/// - `RawWorkerWeights`: Distribute vnodes across workers using the original worker weight values.
+/// - `ActorCounts`: Distribute vnodes based on the number of actors assigned to each worker.
```
Comment (BugenZhao) [inline src/meta/src/stream/stream_graph/assignment.rs:278]: Am I understanding it correctly: if there are a lot of workers/actors in the cluster, the results for different variant should be quite similar in theory, since the number of actors assigned should follow the weights?

File: src/meta/src/stream/stream_graph/assignment.rs [hunk 2]
```diff
@@ -0,0 +1,2127 @@
...
+        let result = assign_items_weighted_with_scale_fn(&containers, &items, 5u8, weighted_scale);
+        let x_count = result.get(&"X").map(Vec::len).unwrap_or(0);
+        let y_count = result.get(&"Y").map(Vec::len).unwrap_or(0);
+        assert_eq!(x_count + y_count, items.len(), "All items must be assigned");
+        assert!(
+            (x_count == 2 && y_count == 1) || (x_count == 1 && y_count == 2),
+            "One container should get 2 items, the other 1, but got {} and {}",
+            x_count,
+            y_count
+        );
```
Comment (BugenZhao) [inline src/meta/src/stream/stream_graph/assignment.rs]: I think the results should be deterministic? Shall we simply demonstrate the expected result in the tests?

File: src/meta/src/stream/stream_graph/assignment.rs (hunk 1 above)
Comment (shanicky) [reply src/meta/src/stream/stream_graph/assignment.rs:278]: Yes, if the cluster is large enough, these two are roughly the same.

It only needs careful consideration when the number of actors is very small but the traffic is very high, to prevent traffic skew.

File: src/meta/src/stream/stream_graph/assignment.rs (hunk 2 above)
Comment (shanicky) [reply src/meta/src/stream/stream_graph/assignment.rs]: Yes, I'll make a change.

**Review by BugenZhao — APPROVED** [review]
Good to see that this new algorithm provides minimal disruption when rescheduling while being stateless!

1. Can you demonstrate which strategy (in the matrix) should be used under different circumstances?
2. The only drawback I can think of is that, vnode assigned to the same actor/worker node is now completely discrete (not continuous). As a result, we cannot effectively leverage the `high_ranges` to merge iterators when performing range scans, and the performance when there's a relatively small amount of data might be degraded.

**Comment by shanicky** [comment]
> Good to see that this new algorithm provides minimal disruption when rescheduling while being stateless!
> 
> 1. Can you demonstrate which strategy (in the matrix) should be used under different circumstances?
> 2. The only drawback I can think of is that, vnode assigned to the same actor/worker node is now completely discrete (not continuous). As a result, we cannot effectively leverage the `high_ranges` to merge iterators when performing range scans, and the performance when there's a relatively small amount of data might be degraded.

I came up with a solution. Is it possible to split vnodes into chunks in certain situations? For example, with 64 vnodes and 16 actors, dividing every 4 vnodes into a vnode chunk would ensure that the vnodes within each actor are contiguous. This should work, but it only guarantees that the vnodes within a worker are not too scattered.

We could even go further by chunking the actors, which might ensure that both the vnodes on the actors and on the workers are ordered.

This way, in some cases, we can trade off some migration affinity (since the basic unit of migration shifts from a vnode to multiple vnodes) to achieve contiguous vnodes.

I'll give it a try.

**Comment by shanicky** [comment]
The vnode chunking mode has already been implemented. You can find more information by checking `VnodeChunkingStrategy`, and you can also refer to the unit test `test_maximize_contiguity_non_divisible_vnodes`. cc @BugenZhao

**Comment by BugenZhao** [comment]
> The vnode chunking mode has already been implemented. You can find more information by checking `VnodeChunkingStrategy`, and you can also refer to the unit test `test_maximize_contiguity_non_divisible_vnodes`. cc @BugenZhao

Cool!

File: src/meta/src/stream/stream_graph/assignment.rs [hunk 3]
```diff
@@ -0,0 +1,2497 @@
...
+///    - If `ActorCounts`: set each worker’s weight = its actor count.
+///    - Run `assign_items_weighted_with_scale_fn` on vnodes vs. the computed weights,
+///      yielding `vnode_to_worker: BTreeMap<W, Vec<V>>`.
+///
+/// 3. **`VNodes` → Actors**
+///    - For each worker, take its vnode list and assign them to actors in simple round-robin:
+///      iterate vnodes in order, dispatching index `% actor_list.len()`.
+///    - Collect into final `BTreeMap<W, BTreeMap<A, Vec<V>>>`.
+pub fn assign_hierarchical<W, A, V, S>(
+    workers: &BTreeMap<W, NonZeroUsize>,
```
Comment (yezizp2012) [inline src/meta/src/stream/stream_graph/assignment.rs:383]: Due to workers now going offline will actively unregister itself, this will cause the worker id to change. In this case, unnecessary vnode migration might occur IIUC. Is it possible to use `advertise addr` + `port` as the parameter?

Comment (yezizp2012) [reply src/meta/src/stream/stream_graph/assignment.rs:383]: Cc @BugenZhao WDYT?

**Comment by shanicky** [comment]
The worker parameters have already been updated to support Clone. 

I plan to merge this PR first.

**Review by yezizp2012 — APPROVED** [review]
